### PR TITLE
feat: Ignore gitignored files when uploading

### PR DESCRIPTION
**Pull request by Aunali321**
Completes #677

### CODE REVIEW

**Comment by pascalbreuninger** [comment]
Hi @Aunali321, thanks for the PR. 
As mentioned in the issue, not uploading `gitignore`d files might be unexpected for some users whilst desired by others. Could you add an option to configure this please? I.e. `devpod up $REPO --gitignore`
If you pass it through the `CLIOptions` struct it will be available in the agent `up` command, and you can pipe it into the grpc client from there

**Comment by Aunali321** [comment]
@pascalbreuninger can you review again? Thanks

**Comment by pascalbreuninger** [comment]
@Aunali321 I think you'll have to vendor the dependencies and fix the golangci-lint errors, currently no pipeline workd

File: cmd/up.go [hunk 1]
```diff
@@ -148,7 +150,15 @@ func (cmd *UpCmd) Run(
 	devPodConfig *config.Config,
 	client client2.BaseWorkspaceClient,
 	log log.Logger,
+	options *CLIOptions,
 ) error {
+	// Create an upCmd object and set its fields based on the CLIOptions
+	up := &upCmd{
```
Comment (pascalbreuninger) [inline cmd/up.go:156]: Not sure where this is being used, is this a leftover?

File: cmd/up.go [hunk 2]
```diff
@@ -54,6 +54,8 @@ type UpCmd struct {
 
 	DotfilesSource string
 	DotfilesScript string
+
+	gitIgnore bool
```
Comment (pascalbreuninger) [inline cmd/up.go:58]: This should be a flag eventually, right? Could you add it to the `upCmd` flags in line 119 and following please?

File: pkg/agent/tunnelserver/stream.go [hunk 3]
```diff
@@ -36,8 +41,22 @@ func NewStreamReader(stream tunnel.Tunnel_StreamWorkspaceClient, log log.Logger)
...
+		gitIgnore = gitignore.NewGitIgnore() // Use an empty IgnoreParser object if parsing fails
+	}
+
+	return &streamWriter{
+		stream:           stream,
+		bytesWritten:     0,
+		lastMessage:      time.Now(),
+		log:              log,
+		gitIgnore:        gitIgnore, // Set the gitIgnore field to the IgnoreParser object
+		gitIgnoreEnabled: gitIgnoreEnabled,
```
Comment (pascalbreuninger) [inline pkg/agent/tunnelserver/stream.go:58]: this field isn't on the struct

File: pkg/agent/tunnelserver/stream.go [hunk 4]
```diff
@@ -36,8 +41,22 @@ func NewStreamReader(stream tunnel.Tunnel_StreamWorkspaceClient, log log.Logger)
...
 }
 
-func NewStreamWriter(stream tunnel.Tunnel_StreamWorkspaceServer, log log.Logger) io.Writer {
-	return &streamWriter{stream: stream, log: log, lastMessage: time.Now()}
+func newStreamWriter(stream tunnel.Tunnel_StreamWorkspaceServer, log log.Logger, gitIgnoreEnabled bool) *streamWriter {
+	// Parse the .gitignore file and create an IgnoreParser object
+	gitIgnore, err := gitignore.CompileIgnoreFile(".gitignore")
+	if err != nil {
+		log.Warnf("Failed to parse .gitignore file: %v", err)
+		gitIgnore = gitignore.NewGitIgnore() // Use an empty IgnoreParser object if parsing fails
```
Comment (pascalbreuninger) [inline pkg/agent/tunnelserver/stream.go:49]: Apparently the package doesn't have this function, what's an alternative here?

File: pkg/agent/tunnelserver/stream.go [hunk 5]
```diff
@@ -46,18 +65,62 @@ type streamWriter struct {
...
+			return nil
+		}
+		relPath, err := filepath.Rel(".", path)
+		if err != nil {
+			return err
+		}
+		if !strings.HasPrefix(relPath, "./") {
+			relPath = "./" + relPath
+		}
+		if !s.gitIgnore.Match(relPath, info.IsDir()) || !s.gitIgnoreEnabled { // Check if gitIgnoreEnabled is true
```
Comment (pascalbreuninger) [inline pkg/agent/tunnelserver/stream.go:87]: Is there a way of doing this check before we start walking the filesystem? Seems like a lot of unnecessary work if we don't want to ignore `.gitignore`

**Review by pascalbreuninger — CHANGES_REQUESTED** [review]
Couple of changes that need to be made, I think you'll need to fix your build and vendor dependencies as well (as mentioned above)

**Comment by Aunali321** [comment]
Sorry i don't have enough Golang language to complete this PR.

**Comment by Aunali321** [comment]
Sorry i don't have enough Golang language to complete this PR.

**Comment by pascalbreuninger** [comment]
Hey @Aunali321, we can still work on that PR if you're interested

**Comment by Aunali321** [comment]
As i said i have 0 knowledge about Golang. i gave it a shot because i had time (which i don't anymore) but thanks!

**Pull request closed without merging**